scripts/globals/additional_effects.lua

```diff
@@ -0,0 +1,263 @@
+function additionalEffectAttack(attacker, defender, baseAttackDamage, item)
+    local addType = item:getMod(tpz.mod.ITEM_ADDEFFECT_TYPE)
+    local subEffect = item:getMod(tpz.mod.ITEM_SUBEFFECT)
+    local damage = item:getMod(tpz.mod.ITEM_ADDEFFECT_DMG)
+    local chance = item:getMod(tpz.mod.ITEM_ADDEFFECT_CHANCE)
+    local element = item:getMod(tpz.mod.ITEM_ADDEFFECT_ELEMENT)
+    local addStatus = item:getMod(tpz.mod.ITEM_ADDEFFECT_STATUS)
+    local power = item:getMod(tpz.mod.ITEM_ADDEFFECT_POWER)
+    local duration = item:getMod(tpz.mod.ITEM_ADDEFFECT_DURATION)
+    local msgID = 0
+    local msgValue = 0
+    local procType =
+    {
+        -- These are arbitrary, make up new ones as needed.
+        NORMAL = 1,
+        HP_HEAL = 2,
+        MP_HEAL = 3,
+        HP_DRAIN = 4,
+        MP_DRAIN = 5,
+        TP_DRAIN = 6,
+        HPMPTP_DRAIN = 7,
+        DISPEL = 8,
+        SELF_BUFF = 9,
+        DEATH = 10,
+    }
+
+    -- If we're not going to proc, lets not execute all those checks!
+    if chance < math.random(100) then
+        return 0,0,0
+    end
+
+    --------------------------------------
+    -- Modifications for proc's sourced from ranged attacks. See notes at top of script.
+    if tpz.addEffect.isRanged(item) then
+        if element then
+            damage = tpz.addEffect.calcRangeBonus(attacker, defender, element, damage)
+        end
+        chance = ltpz.addEffect.evelCorrection(defender:getMainLvl(), attacker:getMainLvl(), chance)
+    end
+    --------------------------------------
+
+    if addType == procType.NORMAL then
+        if addStatus and addStatus > 0 then
+            local tick = tpz.addEffect.statusAttack(addStatus)
+            msgID = tpz.msg.basic.ADD_EFFECT_STATUS
+            defender:addStatusEffect(addStatus, power, tick, duration)
+            msgValue = addStatus
+        end
+
+        if damage > 0 then
+            -- local damage = damage * (math.random(90, 110)/100) -- Artificially forcing 20% variance.
+            damage = tpz.addEffect.calcDamage(attacker, element, defender, damage)
+
+            if subEffect == tpz.subEffect.HP_DRAIN then
+                msgID = tpz.msg.basic.ADD_EFFECT_HP_DRAIN
+                if damage < 0 then
+                    damage = 0
+                end
+            else
+                msgID = tpz.msg.basic.ADD_EFFECT_damage
+                if damage < 0 then
+                    msgID = tpz.msg.basic.ADD_EFFECT_HEAL
+                end
+            end
+            msgValue = damage
+        end
+
+    elseif addType == procType.HP_HEAL then -- Its not a drain and works vs undead. https://www.bg-wiki.com/bg/Dominion_Mace
+        local HP = 10 -- need actual calculation here!
+
+        msgID = tpz.msg.basic.ADD_EFFECT_HP_HEAL
+        attacker:addHP(HP)
+        -- We have to fake this or it will say the defender was healed rather than the attacker.
+        attacker:messageBasic(tpz.msg.basic.ADD_EFFECT_HP_HEAL)
```

Comment:
If `actor` and `target` get switched by `ADD_EFFECT_HP_HEAL` here, won't they also get switched by `ADD_EFFECT_HP_DRAIN` used by drain types?

(Side: We've got an underlying problem wrong if we need to send out messageBasic to workaround action packet implementation)

Reply:
no because the parameters are correct for those messages. And yes we have an underlying problem. An annoying one thats outside the scope of what I am trying to do to get this globalized. This work around is temporary until the core end allows for some control over the parameters (this is a problem with nearly all messages even outside of additional effect - the core **assumes** what the parameters are going to be and edge cases have to be coded around).